### PR TITLE
Small refactor, fix tests

### PR DESCRIPTION
**Pull request by xepozz**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  |

### CODE REVIEW

**Comment by xepozz** [comment]
Test case is here https://github.com/yiisoft/di/pull/167

**Comment by samdark** [comment]
@xepozz would you please resolve conflicts?

**Comment by hiqsol** [comment]
Taken part of this refactoring in https://github.com/yiisoft/di/commit/4042b039f024a384633dac018be7587cead02dd5

File: src/Container.php [hunk 1]
```diff
@@ -51,16 +51,21 @@ public function __construct(
         array $definitions = [],
         array $providers = [],
         ContainerInterface $rootContainer = null
-    ) {
+    )
+    {
         $this->delegateLookup($rootContainer);
+
+        $this->set(ContainerInterface::class, $this->rootContainer ?? $this);
```
Comment (samdark) [inline src/Container.php:58]: Was `has()` check unnecessary? Root container may have it defined.

File: src/Container.php [hunk 2]
```diff
@@ -51,16 +51,21 @@ public function __construct(
...
+        $this->set(Injector::class, new Injector($this->rootContainer ?? $this));
```
Comment (samdark) [inline src/Container.php:59]: Same here. Root container may have it defined.

File: src/Container.php [hunk 3]
```diff
@@ -51,16 +51,21 @@ public function __construct(
...
-        if (!$this->has(ContainerInterface::class)) {
-            $this->set(ContainerInterface::class, $rootContainer ?? $this);
-        }
-        $this->addProviders($providers);
 
+        $this->addProviders($providers);
         # Prevent circular reference to ContainerInterface
-        $this->get(ContainerInterface::class);
+        $container = $this->get(ContainerInterface::class);
+        if ($container !== $this) {
```
Comment (samdark) [inline src/Container.php:66]: Can't we simply check for `$this->rootContainer !== null`?

**Review by samdark — CHANGES_REQUESTED** [review]
Seems it does not pass tests.

File: src/Container.php (hunk 1 above)
Comment (xepozz) [reply src/Container.php:58]: > Root container may have it defined

I don't understand you. Can you please explain what do you mean?

File: src/Container.php (hunk 3 above)
Comment (xepozz) [reply src/Container.php:66]: It's not just a check.
It's combine the fix (early fetching container `$this->get(ContainerInterface::class);`) and redefining `Injector` if container definition was modified.

File: src/Container.php (hunk 1 above)
Comment (samdark) [reply src/Container.php:58]: You have removed a `has()` call. I think it was there to check if there is a definition in a root container already:

```php
$rootContainer = new MyRootContainer();
$rootContainer->define(ContainerInterface::class, function () {
    return new SuperContainer();
});
$yiiContainer = new Container($defintions, $providers, $rootContainer);
```

I am not entirely sure what was the check for. Therefore asking.

**Comment by samdark** [comment]
The idea was mostly included into other PRs so closing. @xepozz thanks for direction.

**Pull request closed without merging**